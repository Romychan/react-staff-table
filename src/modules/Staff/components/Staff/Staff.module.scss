.container {
  @apply max-w-7xl mx-auto py-32;
}

.header {
  @apply flex items-center justify-between mb-16;
}

.title {
  @apply text-3xl font-medium text-gray-900 dark:text-white;
}

.pagination {
  @apply mt-4;
}
