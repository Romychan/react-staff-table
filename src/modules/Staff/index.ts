export { Staff } from './components/Staff/Staff';
