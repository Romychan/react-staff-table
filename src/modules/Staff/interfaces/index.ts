export type { IQueryParams } from './QueryParams';
