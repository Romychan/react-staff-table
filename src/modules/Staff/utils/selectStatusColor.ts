export const selectStatusColor = (state: string) => {
  return state === 'online' ? 'green' : 'default';
};
