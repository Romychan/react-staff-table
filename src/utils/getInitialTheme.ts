import { ThemeType } from '@interfaces/index';

import { DARK_THEME, LIGHT_THEME, THEME_STORAGE_NAME } from '@constants/theme';

export const getInitialTheme = (): ThemeType => {
  const storageTheme = localStorage.getItem(THEME_STORAGE_NAME) as ThemeType;

  if (storageTheme) {
    return storageTheme;
  }

  if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
    return DARK_THEME;
  }

  return LIGHT_THEME;
};
