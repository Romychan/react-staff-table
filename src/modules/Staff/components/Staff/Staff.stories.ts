import type { Meta, StoryObj } from '@storybook/react';

import { usersHandler } from '@tests/server/usersHandler';

import { Staff } from './Staff';

/** This component is used to display staff. */
const meta: Meta<typeof Staff> = {
  title: 'Modules/Staff',
  component: Staff,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  parameters: {
    msw: [usersHandler],
  },
};
