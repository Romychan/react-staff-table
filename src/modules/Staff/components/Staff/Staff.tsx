import { useState, useMemo } from 'react';

import { selectStatusColor } from '../../utils/selectStatusColor';

import { IColumnTable, TableOrderValue, Users } from '@interfaces/index';

import { useFetch } from '@hooks/useFetch';

import { Badge } from '@ui/Badge/Badge';
import { Profile } from '@components/Profile/Profile';
import { Table } from '@components/Table/Table';
import { Pagination } from '@components/Pagination/Pagination';
import { ToggleTheme } from '@components/ToggleTheme/ToggleTheme';
import { StaffHeader } from '../StaffHeader/StaffHeader';

import styles from './Staff.module.scss';

export const Staff = () => {
  const [queryParams, setQueryParams] = useState({
    _sort: 'id' as keyof Users,
    _order: 'asc' as TableOrderValue,
    company_like: '',
    role_like: '',
    _page: '1',
    q: '',
  });

  const columns: IColumnTable<Users, keyof Users>[] = useMemo(
    () => [
      { label: 'ID', accessor: 'id', width: 90, sortable: true },
      {
        label: 'Profile',
        accessor: 'name',
        width: 350,
        sortable: true,
        Cell: ({ currentData }) => (
          <Profile
            name={currentData.name}
            username={currentData.username}
            imageSrc={currentData.avatar}
          />
        ),
      },
      { label: 'Email', accessor: 'email', width: 280 },
      {
        label: 'Status',
        accessor: 'status',
        width: 130,
        sortable: true,
        Cell: ({ currentData }) => (
          <Badge colorScheme={selectStatusColor(currentData.status)}>
            {currentData.status}
          </Badge>
        ),
      },
      { label: 'Role', accessor: 'role', width: 220 },
      { label: 'Company', accessor: 'company', width: 220 },
    ],
    [],
  );

  const {
    data: users,
    isLoading,
    headers,
  } = useFetch<Users[]>(
    `${import.meta.env.VITE_API_URL}/users?${new URLSearchParams(
      queryParams,
    ).toString()}`,
  );

  return (
    <div className={styles.container}>
      <div className={styles.header}>
        <h1 className={styles.title}>Staff Management</h1>
        <ToggleTheme />
      </div>

      <StaffHeader setQueryParams={setQueryParams} />

      <Table
        columns={columns}
        data={users || []}
        isLoading={isLoading}
        sorting={{ column: queryParams._sort, order: queryParams._order }}
        handleSorting={(accessor, order) =>
          setQueryParams((prev) => ({
            ...prev,
            _sort: accessor,
            _order: order,
          }))
        }
      />

      <Pagination
        currentPage={Number(queryParams._page)}
        setCurrentPage={(page) =>
          setQueryParams((prev) => ({ ...prev, _page: page.toString() }))
        }
        totalCount={Number(headers?.get('x-total-count')) || 0}
        className={styles.pagination}
      />
    </div>
  );
};
