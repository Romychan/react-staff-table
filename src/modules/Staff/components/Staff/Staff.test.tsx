import {
  screen,
  render,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import { Staff } from './Staff';

import { server } from '@tests/server';
import { mockMatchMedia } from '@tests/utils/mockMatchMedia';

beforeAll(() => {
  server.listen();
  mockMatchMedia();
});
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('Staff', () => {
  it('should render correctly', async () => {
    render(<Staff />);

    const cellBodyElement = await screen.findByText(/test@test.test/i);

    expect(cellBodyElement).toBeInTheDocument();
  });

  it('should match snapshot', async () => {
    const container = render(<Staff />);

    await waitForElementToBeRemoved(screen.getByTestId('loader'));

    expect(container).toMatchSnapshot();
  });
});
