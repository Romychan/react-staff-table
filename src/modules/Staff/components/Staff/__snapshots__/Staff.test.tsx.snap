// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Staff > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_f6dca3"
      >
        <div
          class="_header_f6dca3"
        >
          <h1
            class="_title_f6dca3"
          >
            Staff Management
          </h1>
          <div
            class="_container_969790"
          >
            <p
              class="_text_969790"
            >
              Light
            </p>
            <input
              class="_input_45a7d5"
              data-testid="switch-toggle"
              type="checkbox"
            />
            <p
              class="_text_969790"
            >
              Dark
            </p>
          </div>
        </div>
        <div
          class="_header_2da5b5"
        >
          <input
            class="_input_76cf42 _input_2da5b5"
            data-testid="input-field"
            placeholder="Search..."
            value=""
          />
          <div
            class="_dropdowns_2da5b5"
          >
            <div
              class="_container_b2c890 _dropdown_2da5b5"
            >
              <div
                class="_dropdown_b2c890"
                data-testid="dropdown-container"
              >
                <button
                  class="_title_b2c890"
                  data-testid="dropdown-header"
                >
                  All Company
                  <svg
                    class="icon icon-chevron _icon_b2c890"
                    data-testid="icon"
                    height="8"
                    stroke=""
                    width="8"
                  >
                    <use
                      xlink:href="/src/assets/img/general/icons.svg#chevron"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              class="_container_b2c890 _dropdown_2da5b5"
            >
              <div
                class="_dropdown_b2c890"
                data-testid="dropdown-container"
              >
                <button
                  class="_title_b2c890"
                  data-testid="dropdown-header"
                >
                  All Role
                  <svg
                    class="icon icon-chevron _icon_b2c890"
                    data-testid="icon"
                    height="8"
                    stroke=""
                    width="8"
                  >
                    <use
                      xlink:href="/src/assets/img/general/icons.svg#chevron"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
        <table
          class="_table_6f0cba"
        >
          <thead
            class="_header_692233"
          >
            <tr
              class="_row_692233"
            >
              <th
                class="_cell_d2877c _sortable_d2877c _asc_d2877c"
                data-testid="table-header-cell"
                style="width: 90px;"
              >
                <div
                  class="_container_d2877c"
                >
                  <svg
                    class="icon icon-arrow _icon_d2877c"
                    data-testid="icon"
                    height="15"
                    stroke=""
                    width="15"
                  >
                    <use
                      xlink:href="/src/assets/img/general/icons.svg#arrow"
                    />
                  </svg>
                  ID
                </div>
              </th>
              <th
                class="_cell_d2877c _sortable_d2877c"
                data-testid="table-header-cell"
                style="width: 350px;"
              >
                <div
                  class="_container_d2877c"
                >
                  <svg
                    class="icon icon-arrow _icon_d2877c"
                    data-testid="icon"
                    height="15"
                    stroke=""
                    width="15"
                  >
                    <use
                      xlink:href="/src/assets/img/general/icons.svg#arrow"
                    />
                  </svg>
                  Profile
                </div>
              </th>
              <th
                class="_cell_d2877c"
                data-testid="table-header-cell"
                style="width: 280px;"
              >
                <div
                  class="_container_d2877c"
                >
                  Email
                </div>
              </th>
              <th
                class="_cell_d2877c _sortable_d2877c"
                data-testid="table-header-cell"
                style="width: 130px;"
              >
                <div
                  class="_container_d2877c"
                >
                  <svg
                    class="icon icon-arrow _icon_d2877c"
                    data-testid="icon"
                    height="15"
                    stroke=""
                    width="15"
                  >
                    <use
                      xlink:href="/src/assets/img/general/icons.svg#arrow"
                    />
                  </svg>
                  Status
                </div>
              </th>
              <th
                class="_cell_d2877c"
                data-testid="table-header-cell"
                style="width: 220px;"
              >
                <div
                  class="_container_d2877c"
                >
                  Role
                </div>
              </th>
              <th
                class="_cell_d2877c"
                data-testid="table-header-cell"
                style="width: 220px;"
              >
                <div
                  class="_container_d2877c"
                >
                  Company
                </div>
              </th>
            </tr>
          </thead>
          <tbody
            class="_body_2d6dd5"
            data-testid="table-body"
          >
            <tr
              class="_row_2d6dd5"
            >
              <td
                class="_cell_2d6dd5"
              >
                1
              </td>
              <td
                class="_cell_2d6dd5"
              >
                <div
                  class="_container_a5316a"
                >
                  <img
                    alt="Gilberte Prin"
                    class="_avatar_3ecbb0"
                    src="./test.png"
                  />
                  <div
                    class="_content_a5316a"
                  >
                    <div
                      class="_name_a5316a"
                    >
                      Gilberte Prin
                    </div>
                    <div
                      class="_username_a5316a"
                    >
                      gprin0
                    </div>
                  </div>
                </div>
              </td>
              <td
                class="_cell_2d6dd5"
              >
                test@test.test
              </td>
              <td
                class="_cell_2d6dd5"
              >
                <span
                  class="_badge_63aabb _green_63aabb"
                >
                  online
                </span>
              </td>
              <td
                class="_cell_2d6dd5"
              >
                Frontend Developer
              </td>
              <td
                class="_cell_2d6dd5"
              >
                Google
              </td>
            </tr>
          </tbody>
        </table>
        <div
          class="_container_08165b _pagination_f6dca3"
          data-testid="pagination"
        >
          <button
            class="_page_08165b _disabled_08165b"
            disabled=""
          >
            Prev
          </button>
          <button
            class="_page_08165b _active_08165b"
            data-id="1"
          >
            1
          </button>
          <button
            class="_page_08165b _disabled_08165b"
            disabled=""
          >
            Next
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_f6dca3"
    >
      <div
        class="_header_f6dca3"
      >
        <h1
          class="_title_f6dca3"
        >
          Staff Management
        </h1>
        <div
          class="_container_969790"
        >
          <p
            class="_text_969790"
          >
            Light
          </p>
          <input
            class="_input_45a7d5"
            data-testid="switch-toggle"
            type="checkbox"
          />
          <p
            class="_text_969790"
          >
            Dark
          </p>
        </div>
      </div>
      <div
        class="_header_2da5b5"
      >
        <input
          class="_input_76cf42 _input_2da5b5"
          data-testid="input-field"
          placeholder="Search..."
          value=""
        />
        <div
          class="_dropdowns_2da5b5"
        >
          <div
            class="_container_b2c890 _dropdown_2da5b5"
          >
            <div
              class="_dropdown_b2c890"
              data-testid="dropdown-container"
            >
              <button
                class="_title_b2c890"
                data-testid="dropdown-header"
              >
                All Company
                <svg
                  class="icon icon-chevron _icon_b2c890"
                  data-testid="icon"
                  height="8"
                  stroke=""
                  width="8"
                >
                  <use
                    xlink:href="/src/assets/img/general/icons.svg#chevron"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="_container_b2c890 _dropdown_2da5b5"
          >
            <div
              class="_dropdown_b2c890"
              data-testid="dropdown-container"
            >
              <button
                class="_title_b2c890"
                data-testid="dropdown-header"
              >
                All Role
                <svg
                  class="icon icon-chevron _icon_b2c890"
                  data-testid="icon"
                  height="8"
                  stroke=""
                  width="8"
                >
                  <use
                    xlink:href="/src/assets/img/general/icons.svg#chevron"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
      <table
        class="_table_6f0cba"
      >
        <thead
          class="_header_692233"
        >
          <tr
            class="_row_692233"
          >
            <th
              class="_cell_d2877c _sortable_d2877c _asc_d2877c"
              data-testid="table-header-cell"
              style="width: 90px;"
            >
              <div
                class="_container_d2877c"
              >
                <svg
                  class="icon icon-arrow _icon_d2877c"
                  data-testid="icon"
                  height="15"
                  stroke=""
                  width="15"
                >
                  <use
                    xlink:href="/src/assets/img/general/icons.svg#arrow"
                  />
                </svg>
                ID
              </div>
            </th>
            <th
              class="_cell_d2877c _sortable_d2877c"
              data-testid="table-header-cell"
              style="width: 350px;"
            >
              <div
                class="_container_d2877c"
              >
                <svg
                  class="icon icon-arrow _icon_d2877c"
                  data-testid="icon"
                  height="15"
                  stroke=""
                  width="15"
                >
                  <use
                    xlink:href="/src/assets/img/general/icons.svg#arrow"
                  />
                </svg>
                Profile
              </div>
            </th>
            <th
              class="_cell_d2877c"
              data-testid="table-header-cell"
              style="width: 280px;"
            >
              <div
                class="_container_d2877c"
              >
                Email
              </div>
            </th>
            <th
              class="_cell_d2877c _sortable_d2877c"
              data-testid="table-header-cell"
              style="width: 130px;"
            >
              <div
                class="_container_d2877c"
              >
                <svg
                  class="icon icon-arrow _icon_d2877c"
                  data-testid="icon"
                  height="15"
                  stroke=""
                  width="15"
                >
                  <use
                    xlink:href="/src/assets/img/general/icons.svg#arrow"
                  />
                </svg>
                Status
              </div>
            </th>
            <th
              class="_cell_d2877c"
              data-testid="table-header-cell"
              style="width: 220px;"
            >
              <div
                class="_container_d2877c"
              >
                Role
              </div>
            </th>
            <th
              class="_cell_d2877c"
              data-testid="table-header-cell"
              style="width: 220px;"
            >
              <div
                class="_container_d2877c"
              >
                Company
              </div>
            </th>
          </tr>
        </thead>
        <tbody
          class="_body_2d6dd5"
          data-testid="table-body"
        >
          <tr
            class="_row_2d6dd5"
          >
            <td
              class="_cell_2d6dd5"
            >
              1
            </td>
            <td
              class="_cell_2d6dd5"
            >
              <div
                class="_container_a5316a"
              >
                <img
                  alt="Gilberte Prin"
                  class="_avatar_3ecbb0"
                  src="./test.png"
                />
                <div
                  class="_content_a5316a"
                >
                  <div
                    class="_name_a5316a"
                  >
                    Gilberte Prin
                  </div>
                  <div
                    class="_username_a5316a"
                  >
                    gprin0
                  </div>
                </div>
              </div>
            </td>
            <td
              class="_cell_2d6dd5"
            >
              test@test.test
            </td>
            <td
              class="_cell_2d6dd5"
            >
              <span
                class="_badge_63aabb _green_63aabb"
              >
                online
              </span>
            </td>
            <td
              class="_cell_2d6dd5"
            >
              Frontend Developer
            </td>
            <td
              class="_cell_2d6dd5"
            >
              Google
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="_container_08165b _pagination_f6dca3"
        data-testid="pagination"
      >
        <button
          class="_page_08165b _disabled_08165b"
          disabled=""
        >
          Prev
        </button>
        <button
          class="_page_08165b _active_08165b"
          data-id="1"
        >
          1
        </button>
        <button
          class="_page_08165b _disabled_08165b"
          disabled=""
        >
          Next
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
